testutil/integration/suites/base_test.go

```diff
@@ -0,0 +1,214 @@
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestGetApp_ReturnsApp() {
+	app := s.NewApp(s.T())
+	require.Same(s.T(), app, s.GetApp())
+}
+func (s *BaseIntegrationSuiteTestSuite) TestSetApp() {
+	// Construct an app.
+	app := s.NewApp(s.T())
+
+	// Reset s.app to nil.
+	s.app = nil
+
+	s.SetApp(app)
+	require.Same(s.T(), app, s.app)
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestGetPoktrollModuleNames() {
+	moduleNames := s.GetPoktrollModuleNames()
+	require.Greater(s.T(), len(moduleNames), 0, "expected non-empty module names")
+	require.ElementsMatch(s.T(), s.poktrollModuleNames, moduleNames)
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestGetCosmosModuleNames() {
+	moduleNames := s.GetCosmosModuleNames()
+	require.Greater(s.T(), len(moduleNames), 0, "expected non-empty module names")
+	require.ElementsMatch(s.T(), s.cosmosModuleNames, moduleNames)
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestSdkCtx() {
+	s.NewApp(s.T())
+	sdkCtx := s.SdkCtx()
+
+	require.NotNil(s.T(), sdkCtx)
+	require.Greater(s.T(), sdkCtx.BlockHeight(), int64(0))
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestFundAddressAndGetBankQueryClient() {
+	s.NewApp(s.T())
+	fundAmount := int64(1000)
+	fundAddr, err := cosmostypes.AccAddressFromBech32(sample.AccAddress())
+	require.NoError(s.T(), err)
+
+	// Assert that the balance is zero before funding.
+	bankQueryClient := s.GetBankQueryClient()
+	balRes, err := bankQueryClient.Balance(s.SdkCtx(), &banktypes.QueryBalanceRequest{
+		Address: fundAddr.String(),
+		Denom:   volatile.DenomuPOKT,
+	})
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), int64(0), balRes.GetBalance().Amount.Int64())
+
+	// Fund the address.
+	s.FundAddress(s.T(), fundAddr, fundAmount)
+
+	// Assert that the balance amount is equal to fundAmount.
+	balRes, err = bankQueryClient.Balance(s.SdkCtx(), &banktypes.QueryBalanceRequest{
+		Address: fundAddr.String(),
+		Denom:   volatile.DenomuPOKT,
+	})
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), fundAmount, balRes.GetBalance().Amount.Int64())
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestFilterLatestEventsWithNewMsgEventMatchFn() {
+	expectedNumEvents := 3
+
+	s.NewApp(s.T())
+	s.emitBankMsgSendEvents(expectedNumEvents)
+
+	// Filter for the "message" type event with "action" attribute value
+	// equal to the MsgSend TypeURL.
+	msgSendTypeURL := cosmostypes.MsgTypeURL(&banktypes.MsgSend{})
+	matchedEvents := s.FilterLatestEvents(events.NewMsgEventMatchFn(msgSendTypeURL))
+
+	require.Equal(s.T(), expectedNumEvents, len(matchedEvents), "unexpected number of matched events")
+
+	// Assert that the events are cleared on the next block.
+	s.GetApp().NextBlock(s.T())
+	require.Equal(s.T(), 0, len(s.SdkCtx().EventManager().Events()), "expected no events in the next block")
+}
+
+func (s *BaseIntegrationSuiteTestSuite) TestFilterLatestEventsWithNewEventTypeMatchFn() {
+	expectedNumEvents := 3
+
+	s.NewApp(s.T())
+	s.emitPoktrollGatewayUnstakedEvents(expectedNumEvents)
```

Comment:
Assert that `len(s.SdkCtx().EventManager().Events())` is zero before emitting and non-zero after, in contrast with The latest assertion in that test.